File: frontend/src/modules/customer/services/customerService.ts
import { apiDelete, apiGet, apiPost, apiPut } from '@/services/apiClient';
import { ApiError, type PageResponse } from '@/shared/types/api';
import type { InvoiceSummaryResponse } from '@/modules/invoice/types';
import type { QuotationSummaryResponse } from '@/modules/quotation/types';
import type {
  CustomerCreditCheckResponse, CustomerFinancialSummaryResponse, CustomerProductHistoryResponse, CustomerRequest,
  CustomerResponse, CustomerSearchParams, CustomerSummaryResponse,
} from '../types';

/** Backend: customer/controller/CustomerController.java */
export const customerService = {
  search: (params: CustomerSearchParams) =>
    apiGet<PageResponse<CustomerSummaryResponse>>('/v1/customers', { params }),

  get: (id: number) => apiGet<CustomerResponse>(`/v1/customers/${id}`),

  create: (body: CustomerRequest) => apiPost<CustomerResponse>('/v1/customers', body),

  update: (id: number, body: CustomerRequest) => apiPut<CustomerResponse>(`/v1/customers/${id}`, body),

  deactivate: (id: number) => apiDelete(`/v1/customers/${id}`),

  financialSummary: (id: number) =>
    apiGet<CustomerFinancialSummaryResponse>(`/v1/customers/${id}/financial-summary`),

  recentInvoices: (id: number, size = 10) =>
    apiGet<PageResponse<InvoiceSummaryResponse>>(`/v1/customers/${id}/invoices`, { params: { size } }),

  recentQuotations: (id: number, size = 10) =>
    apiGet<PageResponse<QuotationSummaryResponse>>(`/v1/customers/${id}/quotations`, { params: { size } }),

  productHistory: (id: number) =>
    apiGet<CustomerProductHistoryResponse[]>(`/v1/customers/${id}/products`),

  /**
   * CR-030 - null means no existing customer at this mobile number (a new
   * walk-in), not a failure. 404 is the server's way of saying exactly
   * that, so it's swallowed here rather than left for every call site to
   * special-case.
   */
  creditCheckByMobile: async (mobile: string): Promise<CustomerCreditCheckResponse | null> => {
    try {
      return await apiGet<CustomerCreditCheckResponse>('/v1/customers/credit-check', { params: { mobile } });
    } catch (error) {
      if (error instanceof ApiError && error.status === 404) return null;
      throw error;
    }
  },
};

File: backend/src/main/java/com/hardware/erp/customer/service/CustomerService.java
package com.hardware.erp.customer.service;

import com.hardware.erp.common.dto.PageResponse;
import com.hardware.erp.customer.dto.CustomerCreditCheckResponse;
import com.hardware.erp.customer.dto.CustomerFinancialSummaryResponse;
import com.hardware.erp.customer.dto.CustomerProductHistoryResponse;
import com.hardware.erp.customer.dto.CustomerRequest;
import com.hardware.erp.customer.dto.CustomerResponse;
import com.hardware.erp.customer.dto.CustomerSummaryResponse;
import com.hardware.erp.customer.entity.CustomerStatus;
import com.hardware.erp.invoice.dto.InvoiceSummaryResponse;
import com.hardware.erp.quotation.dto.QuotationSummaryResponse;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;

public interface CustomerService {

    CustomerResponse create(CustomerRequest request);

    CustomerResponse update(Long id, CustomerRequest request);

    CustomerResponse get(Long id);

    PageResponse<CustomerSummaryResponse> search(String search, CustomerStatus status, Pageable pageable);

    void deactivate(Long id);

    CustomerFinancialSummaryResponse financialSummary(Long id);

    PageResponse<InvoiceSummaryResponse> recentInvoices(Long id, Pageable pageable);

    PageResponse<QuotationSummaryResponse> recentQuotations(Long id, Pageable pageable);

    List<CustomerProductHistoryResponse> productHistory(Long id);

    Optional<CustomerCreditCheckResponse> creditCheckByMobile(String mobileNo);
}
